README.md
# chunked-scatter and scatter-regions

The `chunked-scatter` tool takes a bed file, fasta index, sequence dictionary 
or vcf file as input and divides the
contigs/chromosomes into overlapping chunks of a given size. These chunks will
then be placed in new bed files, one chromosomes per file. Small chromosomes
will be put together to avoid the creation of thousands of files.

The `scatter-regions` tool works in a similar way but with defaults and flags
tuned towards creating genome scatters for GATK tools.

The `safe-scatter` tool produces a more even distribution of sizes in the
output bed files, and guarantees that none of the scatters are smaller than
`--min-input-size`.

## Installation
- Install using pip: `pip install chunked-scatter`
- Install using conda: `conda install chunked-scatter`
    - This requires [conda with a bioconda channel](
    http://bioconda.github.io/user/install.html#).

## Usage

### chunked-scatter
```
usage: chunked-scatter [-h] [-p PREFIX] [-S] [-P] [-c SIZE]
                       [-m MINIMUM_BP_PER_FILE] [-o OVERLAP]
                       INPUT

Given a sequence dict, fasta index or a bed file, scatter over the defined
contigs/regions. Each contig/region will be split into multiple overlapping
regions, which will be written to a new bed file. Each contig will be placed
in a new file, unless the length of the contigs/regions doesn't exceed a given
number.

positional arguments:
  INPUT                 The input file. The format is detected by the
                        extension. Supported extensions are: '.bed', '.dict',
                        '.fai', '.vcf', '.vcf.gz', '.bcf'.

optional arguments:
  -h, --help            show this help message and exit
  -p PREFIX, --prefix PREFIX
                        The prefix of the ouput files. Output will be named
                        like: <PREFIX><N>.bed, in which N is an incrementing
                        number. Default 'scatter-'.
  -S, --split-contigs   If set, contigs are allowed to be split up over
                        multiple files.
  -P, --print-paths     If set prints paths of the output files to STDOUT.
                        This makes the program usable in scripts and
                        worfklows.
  -c SIZE, --chunk-size SIZE
                        The size of the chunks. The first chunk in a region or
                        contig will be exactly length SIZE, subsequent chunks
                        will SIZE + OVERLAP and the final chunk may be
                        anywhere from 0.5 to 1.5 times SIZE plus overlap. If a
                        region (or contig) is smaller than SIZE the original
                        regions will be returned. Defaults to 1e6
  -m MINIMUM_BP_PER_FILE, --minimum-bp-per-file MINIMUM_BP_PER_FILE
                        The minimum number of bases represented within a
                        single output bed file. If an input contig or region
                        is smaller than this MINIMUM_BP_PER_FILE, then the
                        next contigs/regions will be placed in the same file
                        untill this minimum is met. Defaults to 45e6.
  -o OVERLAP, --overlap OVERLAP
                        The number of bases which each chunk should overlap
                        with the preceding one. Defaults to 150.
```

### scatter-regions
```
usage: scatter-regions [-h] [-p PREFIX] [-S] [-P] [-s SCATTER_SIZE] INPUT

Given a sequence dict, fasta index or a bed file, scatter over the defined
contigs/regions. Creates a bed file where the contigs add up approximately to
the given scatter size.

positional arguments:
  INPUT                 The input file. The format is detected by the
                        extension. Supported extensions are: '.bed', '.dict',
                        '.fai', '.vcf', '.vcf.gz', '.bcf'.

optional arguments:
  -h, --help            show this help message and exit
  -p PREFIX, --prefix PREFIX
                        The prefix of the ouput files. Output will be named
                        like: <PREFIX><N>.bed, in which N is an incrementing
                        number. Default 'scatter-'.
  -S, --split-contigs   If set, contigs are allowed to be split up over
                        multiple files.
  -P, --print-paths     If set prints paths of the output files to STDOUT.
                        This makes the program usable in scripts and
                        worfklows.
  -s SCATTER_SIZE, --scatter-size SCATTER_SIZE
                        The maximum size for the regions over which to
                        scatter. If contigs are not split, and a contig is
                        bigger than the maximum size, the contig will be
                        placed in its own file. Default: 1000000000.
```

### safe-scatter
```
usage: safe-scatter [-h] [-p PREFIX] [-P] [-c SCATTER_COUNT]
                    [-m MIN_SCATTER_SIZE] [--mix-small-regions]
                    INPUT

Given a sequence dict, fasta index or a bed file, scatter over the defined
contigs/regions. Creates a bed file where the contigs add up to the average
scatter size to within min_scatter_size. NOTE, this tool always splits up
contigs.

positional arguments:
  INPUT                 The input file. The format is detected by the
                        extension. Supported extensions are: '.bed', '.dict',
                        '.fai', '.vcf', '.vcf.gz', '.bcf'.

optional arguments:
  -h, --help            show this help message and exit
  -p PREFIX, --prefix PREFIX
                        The prefix of the ouput files. Output will be named
                        like: <PREFIX><N>.bed, in which N is an incrementing
                        number. Default 'scatter-'. (default: scatter-)
  -P, --print-paths     If set prints paths of the output files to STDOUT.
                        This makes the program usable in scripts and
                        worfklows. (default: False)
  -c SCATTER_COUNT, --scatter-count SCATTER_COUNT
                        The number of chunks to scatter the regions in. All
                        chunks will be within --min-scatter-size of each other
                        except for the final chunk. (default: 50)
  -m MIN_SCATTER_SIZE, --min-scatter-size MIN_SCATTER_SIZE
                        The minimum size of a scatter. This tool will never
                        generate regions smaller than this value, unless the
                        original regions aresmaller. (default: 10000)
  --mix-small-regions   Mix small regions with regular regions in the input
                        regions. This can be useful in case there is a bias in
                        the composition of the regions. For example, the human
                        reference genome has all unplaced contigs (which are
                        small and difficult to process) at the end of the
                        file, which means they all end up in the same bedfile.
                        Enabling mixing prevents this (default: False)
```

## Examples
### bed file
Given a bed file located at `/data/regions.bed`:
```
chr1	100	1000
chr1	2000	16000
chr2	5000	10000
```

The command:
```
chunked-scatter -p /data/scatter_ -m 1000 -c 5000 /data/regions.bed
```

Will produce the following two output files:
- `/data/scatter_0.bed`:
  ```
  chr1	100	1000
  chr1	2000	7000
  chr1	6850	12000
  chr1	11850	16000
  ```
- `/data/scatter_1.bed`:
  ```
  chr2	5000	10000
  ```

### dict file
Given a dict file located at `/data/ref.dict`:
```
@SQ	SN:chr1	LN:3000000
@SQ SN:chr2 LN:500000
```

The command:
```
chunked-scatter -p /data/scatter_ /data/regions.bed
```

Will produce the following output file at `/data/scatter_0.bed`:
```
chr1	0	1000000
chr1	999850	2000000
chr1	1999850	3000000
chr2	0	500000
```


CHANGELOG.rst
==========
Changelog
==========

version 1.1.0
---------------------------
+ The ``--scatter-count`` option is now required.
+ Added option ``--mix-small-regions`` to evenly mix small regions in the input.
+ Added ``safe-scatter`` which produces a more even scattering.

version 1.0.0
---------------------------
+ VCF files can now also be used as input files. `Pysam
  <https://github.com/pysam-developers/pysam>`_ was used to implement this
  functionality.
+ scatter-regions now uses maximum size instead of minimum size for its scatter
  sizes. This leads to a more even distribution.

version 0.2.0
---------------------------
+ The ``--input`` has now been abolished in favor of a positional input
  argument.
+ The ``--prefix`` flag now has ``scatter-`` as default value.
+ Added a ``--print-paths`` or ``-P`` flag that prints the paths of the output
  files to stdout.
+ Added support for fasta index (``.fai``) files.
+ Added a ``--split-contigs`` flag to allow contigs to be split up over
  multiple files.
+ Added new ``scatter-regions`` command that is a replacement for
  `biopet-scatterregions <https://github.com/biopet/scatterregions>`_. Unlike
  biopet-scatterregions, ``scatter-regions`` works properly with scatter sizes
  larger than the Java's integer limit (2147483647). Scatter-regions invokes
  the same algorithms as ``chunked-scatter`` but with defaults that make more
  sense for a GATK scattering use case.
+ The code has been refactored for better maintainability. Python 3.5 is no
  longer supported.

version 0.1.0
---------------------------
+ First release of chunked scatter


tests/test_main.py
# Copyright (c) 2019 Leiden University Medical Center
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

import sys
from pathlib import Path

from chunked_scatter.chunked_scatter import main, parse_args
from chunked_scatter.safe_scatter import main as safe_scatter_main
from chunked_scatter.scatter_regions import main as scatter_regions_main

DATA_DIR = Path(__file__).parent / Path("data")


def test_bed_input(tmpdir, capsys):
    sys.argv = ["script", "-p", "{}/test_result_".format(tmpdir),
                str(Path(DATA_DIR, "regions.bed")), "-c", "5000", "-m", "1",
                "-P"]
    main()
    file1 = tmpdir / Path("test_result_0.bed")
    file2 = tmpdir / Path("test_result_1.bed")
    expected_file1 = ("chr1\t100\t1000\n"
                      "chr1\t2000\t7000\n"
                      "chr1\t6850\t12000\n"
                      "chr1\t11850\t16000\n")
    expected_file2 = "chr2\t5000\t10000\n"
    print(file1, file1.exists())
    print(file2, file2.exists())
    assert file1.exists()
    assert file2.exists()
    assert file1.read() == expected_file1
    assert file2.read() == expected_file2
    assert str(tmpdir / Path("test_result_0.bed")) in capsys.readouterr().out


def test_dict_input(tmpdir):
    sys.argv = ["script", "-p", "{}/test_result_".format(tmpdir),
                str(Path(DATA_DIR, "ref.dict"))]
    main()
    file1 = tmpdir / Path("test_result_0.bed")
    expected_file1 = ("chr1\t0\t1000000\n"
                      "chr1\t999850\t2000000\n"
                      "chr1\t1999850\t3000000\n"
                      "chr2\t0\t500000\n")
    print(file1, file1.exists())
    assert file1.exists()
    assert file1.read() == expected_file1


def test_parse_args():
    sys.argv = ["script", "input.bed"]
    args = parse_args()
    assert args.input == "input.bed"
    assert args.prefix == "scatter-"
    assert args.chunk_size == 1e6
    assert args.minimum_bp_per_file == 45e6
    assert args.overlap == 150


def test_scatter_regions_main(tmpdir, capsys):
    sys.argv = ["scatter-regions", "-p",
                str(Path(str(tmpdir), "scatters", "scatter-")),
                "--split-contigs",
                "-s", "1100000",
                str(Path(DATA_DIR, "ref.dict")),
                "--print-paths"]
    scatter_regions_main()
    assert Path(str(tmpdir), "scatters", "scatter-0.bed").exists()
    assert Path(str(tmpdir), "scatters", "scatter-1.bed").exists()
    assert Path(str(tmpdir), "scatters", "scatter-2.bed").exists()
    assert Path(str(tmpdir), "scatters", "scatter-3.bed").exists()
    assert Path(str(tmpdir), "scatters", "scatter-2.bed").read_text() == (
        "chr1\t2200000\t3000000\n"
    )
    captured = capsys.readouterr()
    assert str(Path(str(tmpdir), "scatters", "scatter-0.bed")) in captured.out


def test_safe_scatter_main(tmpdir, capsys):
    sys.argv = ["safe-scatter", "-p",
                str(Path(str(tmpdir), "scatters", "scatter-")),
                "--scatter-count", "3",
                str(Path(DATA_DIR, "ref.dict")),
                "--print-paths"]
    safe_scatter_main()
    assert Path(str(tmpdir), "scatters", "scatter-0.bed").exists()
    assert Path(str(tmpdir), "scatters", "scatter-1.bed").exists()
    assert Path(str(tmpdir), "scatters", "scatter-2.bed").exists()
    assert not Path(str(tmpdir), "scatters", "scatter-3.bed").exists()
    assert Path(str(tmpdir), "scatters", "scatter-1.bed").read_text() == (
        "chr1\t1160000\t2320000\n"
    )
    captured = capsys.readouterr()
    assert str(Path(str(tmpdir), "scatters", "scatter-0.bed")) in captured.out


tests/test_parsers.py
# Copyright (c) 2019 Leiden University Medical Center
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

from pathlib import Path

from chunked_scatter.parsers import BedRegion, file_to_regions

import pytest

datadir = Path(__file__).parent / Path("data")


def test_file_to_regions_bed():
    result = list(file_to_regions(datadir / "regions.bed"))
    assert result == [
        BedRegion("chr1", 100, 1000),
        BedRegion("chr1", 2000, 16000),
        BedRegion("chr2", 5000, 10000)
    ]


def test_file_to_regions_dict():
    result = list(file_to_regions(datadir / "ref.dict"))
    assert result == [
        BedRegion("chr1", 0, 3000000),
        BedRegion("chr2", 0, 500000)
    ]


def test_file_to_regions_fai():
    result = list(file_to_regions(datadir / "reference.fasta.fai"))
    assert result == [
        BedRegion("chr1", 0, 15624),
        BedRegion("chrM", 0, 16569),
        BedRegion("chr1_alt", 0, 6120)
    ]


def test_file_to_regions_vcf():
    result = list(file_to_regions(datadir / "example.vcf"))
    assert result == [
        BedRegion("22", 499, 500),
        BedRegion("22", 999, 1000),
        BedRegion("22", 1001, 1002),
        BedRegion("22", 1099, 1101)
    ]


def test_file_to_regions_wrong_ext(capsys):
    with pytest.raises(NotImplementedError) as error:
        file_to_regions("input")
    error.match("Unkown extension '' for file: 'input'. Supported extensions "
                "are:")


tests/test_chunkers.py
# Copyright (c) 2019 Leiden University Medical Center
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.
from chunked_scatter.chunked_scatter import BedRegion, chunked_scatter, \
    region_chunker

import pytest

BED_REGIONS = [BedRegion("chr1", 100, 1000), BedRegion("chr1", 2000, 16000),
               BedRegion("chr2", 5000, 10000)]

DICT_REGIONS = [BedRegion("chr1", 0, 3000000), BedRegion("chr2", 0, 500000)]

REGION_TESTS = [
    (BED_REGIONS, 5000, 150, [
        BedRegion("chr1", 100, 1000),
        BedRegion("chr1", 2000, 7000),
        BedRegion("chr1", 6850, 12000),
        BedRegion("chr1", 11850, 16000),
        BedRegion("chr2", 5000, 10000)
    ]),
    (BED_REGIONS, 5000, 0, [
        BedRegion("chr1", 100, 1000),
        BedRegion("chr1", 2000, 7000),
        BedRegion("chr1", 7000, 12000),
        BedRegion("chr1", 12000, 16000),
        BedRegion("chr2", 5000, 10000)
    ]),
    (DICT_REGIONS, 1_000_000, 150, [
        BedRegion("chr1", 0, 1_000_000),
        BedRegion("chr1", 999850, 2_000_000),
        BedRegion("chr1", 1999850, 3_000_000),
        BedRegion("chr2", 0, 500_000)
    ]),
    (DICT_REGIONS, 1_000_000, 0, [
        BedRegion("chr1", 0, 1_000_000),
        BedRegion("chr1", 1_000_000, 2_000_000),
        BedRegion("chr1", 2_000_000, 3_000_000),
        BedRegion("chr2", 0, 500_000)
    ]),
    (DICT_REGIONS, 10**12, 150, [
        BedRegion("chr1", 0, 3_000_000),
        BedRegion("chr2", 0, 500_000)
    ])
]


@pytest.mark.parametrize(["regions", "chunk_size", "overlap", "result"],
                         REGION_TESTS)
def test_region_chunker(regions, chunk_size, overlap, result):
    chunks = list(region_chunker(regions, chunk_size, overlap))
    assert chunks == result


def test_chunked_scatter():
    result = list(chunked_scatter(BED_REGIONS, 5000, 150, 1))
    assert result == [
        [BedRegion("chr1", 100, 1000),
         BedRegion("chr1", 2000, 7000),
         BedRegion("chr1", 6850, 12000),
         BedRegion("chr1", 11850, 16000)
         ],
        [BedRegion("chr2", 5000, 10000)],
    ]


def test_chunked_scatter_split_contigs():
    result = list(chunked_scatter(BED_REGIONS, 5000, 150, 10000,
                                  contigs_can_be_split=True))
    assert result == [
        [BedRegion("chr1", 100, 1000),
         BedRegion("chr1", 2000, 7000),
         BedRegion("chr1", 6850, 12000)],
        [BedRegion("chr1", 11850, 16000),
         BedRegion("chr2", 5000, 10000)],
    ]


src/chunked_scatter/parsers.py
# Copyright (c) 2019 Leiden University Medical Center
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

import os
from typing import Generator, NamedTuple, Optional, Union

from pysam import VariantFile, VariantRecord

# Add extensions here so they can be used troughout the project for messages.
SUPPORTED_EXTENSIONS = [".bed", ".dict", ".fai", ".vcf", ".vcf.gz", ".bcf"]
SUPPORTED_EXTENSIONS_STRING = "'" + "', '".join(SUPPORTED_EXTENSIONS) + "'"


class BedRegion(NamedTuple):
    """A class that contains a region described as in the BED file format."""
    contig: str
    start: int
    end: int

    def __str__(self):
        return f"{self.contig}\t{self.start}\t{self.end}"

    def __len__(self):
        return self.end - self.start


def dict_file_to_regions(in_file: Union[str, os.PathLike]
                         ) -> Generator[BedRegion, None, None]:
    """
    Converts a Picard SequenceDictionary file to a BedRegion Generator.
    :param in_file: The sequence dictionary
    :return: A generator of BedRegions
    """
    with open(in_file, "rt") as in_file_h:
        for line in in_file_h:
            fields = line.strip().split()
            if fields[0] != "@SQ":
                continue

            contig: Optional[str] = None
            length: Optional[int] = None
            for field in fields:
                if field.startswith("LN"):
                    length = int(field[3:])
                elif field.startswith("SN"):
                    contig = field[3:]
            if contig and length:
                yield BedRegion(contig, 0, length)


def bed_file_to_regions(in_file: Union[str, os.PathLike]
                        ) -> Generator[BedRegion, None, None]:
    """
    Converts a BED file to a generator of BED regions
    :param in_file: The BED file
    :return: A BedRegion Generator
    """
    with open(in_file, "rt") as in_file_h:
        for line in in_file_h:
            fields = line.strip().split()
            # Skip browser and track fields and other invalid lines.
            if fields[0] in ["browser", "track"] or len(fields) < 3:
                continue
            # Take the first 3 columns of each line to create a new BedRegion
            yield BedRegion(fields[0], int(fields[1]), int(fields[2]))


def fai_file_to_regions(in_file: Union[str, os.PathLike]
                        ) -> Generator[BedRegion, None, None]:
    # faidx format described here: https://www.htslib.org/doc/faidx.html
    with open(in_file, "rt") as in_file_h:
        for line in in_file_h:
            # faidx has name, length, offset, linebases, linewidth columns. And
            # optionally a qualoffset. By using maxsplit=2, we catch name and
            # length, while not attempting to distinguish whether we have 5 or
            # 6 columns.
            name, length, _ = line.strip().split(maxsplit=2)
            yield BedRegion(name, 0, int(length))


def vcf_file_to_regions(in_file: Union[str, os.PathLike]):
    vcf = VariantFile(in_file, mode="r")
    try:  # VariantFile automatically opens file
        for variant in vcf:  # type: VariantRecord
            yield BedRegion(variant.contig, variant.start, variant.stop)
    finally:
        # Make sure vcf is always closed
        vcf.close()


def file_to_regions(in_file: Union[str, os.PathLike]):
    base, extension = os.path.splitext(in_file)
    if extension == ".bed":
        return bed_file_to_regions(in_file)
    elif extension == ".dict":
        return dict_file_to_regions(in_file)
    elif extension == ".fai":
        return fai_file_to_regions(in_file)
    elif extension in (".vcf", ".bcf", ".vcf.gz"):
        return vcf_file_to_regions(in_file)
    else:
        raise NotImplementedError(
            f"Unkown extension '{extension}' for file: '{in_file}'. Supported "
            f"extensions are: {SUPPORTED_EXTENSIONS_STRING}.")


src/chunked_scatter/chunked_scatter.py
# Copyright (c) 2019 Leiden University Medical Center
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

import argparse
from pathlib import Path
from typing import Generator, Iterable, List

from .parsers import BedRegion, SUPPORTED_EXTENSIONS_STRING, file_to_regions


def region_chunker(regions: Iterable[BedRegion], chunk_size: int, overlap: int
                   ) -> Generator[BedRegion, None, None]:
    """
    Converts each region into chunks if the chunk_size is smaller than the
    region size.
    :param regions: The regions which to chunk.
    :param chunk_size: The size of the chunks.
    :param overlap: The size of the overlap between chunks.
    :return: The new chunked regions.
    """
    for contig, start, end in regions:
        position = start
        # This will cause the last chunk to be between 0.5 and 1.5
        # times the chunk_size in length, this way we avoid the
        # possibility that the last chunk ends up being to small
        # (eg. 1+overlap bases).
        while position + chunk_size * 1.5 < end:
            if position - overlap <= start:
                yield BedRegion(contig, start, int(position + chunk_size))
            else:
                yield BedRegion(contig, int(position - overlap),
                                int(position + chunk_size))
            position += chunk_size
        if position - overlap <= start:
            yield BedRegion(contig, start, end)
        else:
            yield BedRegion(contig, int(position - overlap), end)


def chunked_scatter(regions: Iterable[BedRegion],
                    chunk_size: int,
                    overlap: int,
                    list_size: int,
                    size_is_maximum: bool = False,
                    contigs_can_be_split: bool = False,
                    ) -> Generator[List[BedRegion], None, None]:
    """
    Scatter regions in chunks with an overlap. It returns Lists of regions
    where each list of regions has the regions describe at least the amount
    of base pairs in minimum bas pairs. Except the last list.
    :param regions: The regions which to chunk.
    :param chunk_size: The size of each chunk.
    :param overlap: How much overlap there should be between chunks.
    :param list_size: What the minimum amount of base pairs should be
    that the regions encompass per List.
    :param size_is_maximum: Use list_size as a maximum instead of a minimum
    :param contigs_can_be_split: Whether contigs (chr1, for example) are
    allowed to be split across multiple lists.
    :return: Lists of BedRegions, which can be converted into BED files.
    """
    current_scatter_size = 0
    current_contig = None
    chunk_list: List[BedRegion] = []
    for chunk in region_chunker(regions, chunk_size, overlap):
        # If the next chunk is on a different contig
        if contigs_can_be_split or chunk.contig != current_contig:
            current_contig = chunk.contig
            # Yield if the minimum is reached, or if current chunk will
            # overflow the size and there is at least one chunk already.
            size_to_check = (current_scatter_size + len(chunk)
                             if size_is_maximum else current_scatter_size)
            if size_to_check >= list_size and len(chunk_list) > 0:
                yield chunk_list
                chunk_list = []
                current_scatter_size = 0
        # Add the chunk to the bed file
        chunk_list.append(chunk)
        current_scatter_size += len(chunk)
    # If there are leftovers yield them.
    if chunk_list:
        yield chunk_list


def region_lists_to_scatter_files(region_lists: Iterable[List[BedRegion]],
                                  prefix: str) -> List[str]:
    """
    Convert lists of BedRegions to '{prefix}{number}.bed' files. The number
    starts at 0 and is increased with 1 for each file.
    :param region_lists: The region lists to be converted into BED files.
    :param prefix: The filename prefix for the BedFiles
    :return: A list of filenames of the written paths.
    """
    parent_dir = Path(prefix).parent
    if not parent_dir.exists():
        parent_dir.mkdir(parents=True)
    output_files: List[str] = []
    for scatter_number, region_list in enumerate(region_lists):
        out_file = f"{prefix}{scatter_number}.bed"
        with open(out_file, "wt") as out_file_h:
            for bed_region in region_list:
                out_file_h.write(str(bed_region) + '\n')
        # I much prefer yield out_file instead. But this means the function
        # won't do anything until it is iterated over, which is not nice.
        output_files.append(out_file)
    return output_files


def common_parser() -> argparse.ArgumentParser:
    """Commmon arguments for chunked-scatter and scatter-regions."""
    parser = argparse.ArgumentParser()
    parser.add_argument("-p", "--prefix", type=str, default="scatter-",
                        help="The prefix of the ouput files. Output will be "
                        "named like: <PREFIX><N>.bed, in which N is an "
                        "incrementing number. Default 'scatter-'.")
    parser.add_argument("input", metavar="INPUT", type=str,
                        help=f"The input file. The format is detected by the "
                             f"extension. Supported extensions are: "
                             f"{SUPPORTED_EXTENSIONS_STRING}.")
    parser.add_argument("-P", "--print-paths", action="store_true",
                        help="If set prints paths of the output files to "
                             "STDOUT. This makes the program usable in "
                             "scripts and worfklows.")
    return parser


def parse_args():
    """Argument parser for the chunked-scatter program."""
    parser = common_parser()
    parser.description = (
        "Given a sequence dict, fasta index or a bed file, scatter over the "
        "defined contigs/regions. Each contig/region will be split into "
        "multiple overlapping regions, which will be written to a new bed "
        "file. Each contig will be placed in a new file, unless the length of "
        "the contigs/regions doesn't exceed a given number.")

    parser.add_argument("-c", "--chunk-size", type=int, default=1e6,
                        metavar="SIZE",
                        help="The size of the chunks. The first chunk in a "
                        "region or contig will be exactly length SIZE, "
                        "subsequent chunks will SIZE + OVERLAP and the final "
                        "chunk may be anywhere from 0.5 to 1.5 times SIZE "
                        "plus overlap. If a region (or contig) is smaller "
                        "than SIZE the original regions will be returned. "
                        "Defaults to 1e6")
    parser.add_argument("-m", "--minimum-bp-per-file", type=int, default=45e6,
                        help="The minimum number of bases represented within "
                        "a single output bed file. If an input contig or "
                        "region is smaller than this MINIMUM_BP_PER_FILE, "
                        "then the next contigs/regions will be placed in the "
                        "same file untill this minimum is met. Defaults to "
                        "45e6.")
    parser.add_argument("-o", "--overlap", type=int, default=150,
                        help="The number of bases which each chunk should "
                        "overlap with the preceding one. Defaults to 150.")
    parser.add_argument("-S", "--split-contigs", action="store_true",
                        help="If set, contigs are allowed to be split up over "
                             "multiple files.")
    args = parser.parse_args()
    return args


def main():
    args = parse_args()
    scattered_chunks = chunked_scatter(file_to_regions(args.input),
                                       args.chunk_size, args.overlap,
                                       args.minimum_bp_per_file,
                                       size_is_maximum=False,
                                       contigs_can_be_split=args.split_contigs)
    out_files = region_lists_to_scatter_files(scattered_chunks, args.prefix)
    if args.print_paths:
        print("\n".join(out_files))


if __name__ == "__main__":  # pragma: no cover
    main()
